### ISSUE TITLE
nan for focal loss when training ssd

### ISSUE THREAD

**Issue by yzl96** [open-mmlab/mmdetection#1245]
Hi, I want to use focal loss to train ssd, and I change the ssd code, but the loss is always nan.
the file I change is below:
in ssd_head.py:
`    def loss_single(self, cls_score, bbox_pred, labels, label_weights,
                    bbox_targets, bbox_weights, num_total_samples, cfg):
        # loss_cls_all = F.cross_entropy(
        #     cls_score, labels, reduction='none') * label_weights
        # pos_inds = (labels > 0).nonzero().view(-1)
        # neg_inds = (labels == 0).nonzero().view(-1)
        # print("just to see anchor's imbalance: ")
        # print("positive anchors: ")
        # print(pos_inds.size(0))
        # print(neg_inds.size(0))
        # print("negtive relative to positive:")
        # print(neg_inds.size(0) / pos_inds.size(0))
        # num_pos_samples = pos_inds.size(0)
        # num_neg_samples = cfg.neg_pos_ratio * num_pos_samples
        # if num_neg_samples > neg_inds.size(0):
        #     num_neg_samples = neg_inds.size(0)
        # topk_loss_cls_neg, _ = loss_cls_all[neg_inds].topk(num_neg_samples)
        # loss_cls_pos = loss_cls_all[pos_inds].sum()
        # loss_cls_neg = topk_loss_cls_neg.sum()
        # loss_cls = (loss_cls_pos + loss_cls_neg) / num_total_samples

        labels = torch.tensor(labels, dtype=torch.long)
        labels = torch.nn.functional.one_hot(labels, num_classes=21).cuda()

        loss_cls = py_sigmoid_focal_loss(
            cls_score, labels, label_weights, avg_factor=num_total_samples)

        loss_bbox = smooth_l1_loss(
            bbox_pred,
            bbox_targets,
            bbox_weights,
            beta=cfg.smoothl1_beta,
            avg_factor=num_total_samples)
        return loss_cls, loss_bbox #loss_cls[None]`
I comment the entropy cross loss, and change to focal loss just like above.
before enter focal loss, I convert label to one-hot format.
The training output:
![Screenshot from 2019-08-23 16-46-29](https://user-images.githubusercontent.com/31734981/63579588-97f29f00-c5c5-11e9-9b99-796bf40ca40e.png)

I search on the internet, and I try : pred_sigmoid = pred_sigmoid.clamp(min=0.0001, max=1.0) in py_sigmoid_focal_loss , but it's no use.

**Comment by hellock** [comment]
SSD and RetinaNet have different sampling strategies, and directly applying focal loss to SSD is very likely to fail, even though we have not tried that. You may need to careful tuning on the parameters and even the pipeline to make it work.

**Comment by yzl96** [comment]
@hellock ,I change the loss follow the step:
1. find where ssd compute classification loss (single_loss function)
2. comment the CE loss
3. find how retinanet use focal loss(config loss, and  call build_loss(loss_cls_config), then call self.cls_loss in single_loss function)
4. follow retinanet , add focal loss to ssd

 the classification loss of ssd is defined in 
https://github.com/open-mmlab/mmdetection/blob/1d79715320f5fa61a7a52334594b46c261f55865/mmdet/models/anchor_heads/ssd_head.py#L111-L125
, I follow retinanet to use focal loss, I add those code in ssd_head.py:
`loss_cls_config = dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0)

        self.loss_cls = build_loss(loss_cls_config)
        in ssd_head.py's init function.     
`
`loss_cls = self.loss_cls(cls_score, labels, label_weights, avg_factor=num_total_samples)
in loss_single function`

I already train with CE loss 240 epochs, if I train ssd with focal loss from 100th epoch checkpoint or from start , loss quickly go nan, but if I train ssd with focal loss from 240th epoch, the loss don't go nan, I don't think my change has any problem, for focal loss, I use default parameters.

**Comment by helloworld77** [comment]
Focalloss focus on class imbalance of detection problems. So, when I use focalloss in ssd, I will ban the max negative mine catalog in original ssd, and make sure most positive and negative objects are calculated in focalloss.

**Comment by yzl96** [comment]
> Focalloss focus on class imbalance of detection problems. So, when I use focalloss in ssd, I will ban the max negative mine catalog in original ssd, and make sure most positive and negative objects are calculated in focalloss.

Thanks you, I have solve the problem, after I limit the gradient, the loss can drop, but the performance of focal loss is lower than 1:3 loss, I did not figure why.

**Comment by jeevan-thapa** [comment]
> SSD and RetinaNet have different sampling strategies, and directly applying focal loss to SSD is very likely to fail, even though we have not tried that. You may need to careful tuning on the parameters and even the pipeline to make it work.

Do single-stage detectors too require Sampling?

**Comment by zhaoxin111** [comment]
> > Focalloss focus on class imbalance of detection problems. So, when I use focalloss in ssd, I will ban the max negative mine catalog in original ssd, and make sure most positive and negative objects are calculated in focalloss.
> 
> Thanks you, I have solve the problem, after I limit the gradient, the loss can drop, but the performance of focal loss is lower than 1:3 loss, I did not figure why.

Have you solve the problem. I have similar problem that the loss is very  large.

**Comment by leewi9** [comment]
@yzl96 hello, I have the same problem, can you note here how to modify all the codes? thanks

**Comment by kingstarcraft** [comment]
I think it is the bug for getting num_total_sample at  line 467-468 of anchor_head.py  when using focal_loss
I set use_all_sample is true when use focal_loss at init function of AnchorHead
```python
        self.sampling = loss_cls['type'] not in [
            'FocalLoss', 'GHMC', 'QualityFocalLoss'
        ]
        self.use_all_sample = loss_cls['type'] == 'FocalLoss'
        if self.use_sigmoid_cls:
            self.cls_out_channels = num_classes
        else:
            self.cls_out_channels = num_classes + 1
```
and modify the code at line 467-468
```python
        num_total_samples = (
            num_total_pos + num_total_neg if (self.sampling or self.use_sigmoid_cls) else num_total_pos)
```
